### ISSUE TITLE
Blank output when using External Markdown in reveal.js

### ISSUE THREAD

**Issue by vernhart** [astefanutti/decktape#12]
To test this, I started with the sample html file that comes with reveal.js and stripped out the entire `<div class="reveal"> ... </div>` and replaced it with the following code bits.

This works:

    <div class="reveal">
      <div class="slides">
        <section>Normal Text</section>
      </div>
    </div>

This works:

    <div class="reveal">
      <div class="slides">
        <section data-markdown>
          <script type="text/template">
            ## Markdown Text
          </script>
        </section>
      </div>
    </div>

But if I put "## Markdown Text" inside a file called text.md and use the following in my html, it looks exactly as the above test when I view it in my browser but I get a blank pdf from decktape.

    <div class="reveal">
      <div class="slides">
        <section data-markdown="text.md"/>
      </div>
    </div>

Any thoughts?

**Comment by astefanutti** [comment]
Indeed, I've been able to reproduce this issue with the reveal.js Markdown example: https://github.com/hakimel/reveal.js/blob/44548ba357490943cfa6a2e5139863bd4fc4f66a/plugin/markdown/example.html.

This is likely to be dealing with the loading of the external markdown document: https://github.com/hakimel/reveal.js/blob/44548ba357490943cfa6a2e5139863bd4fc4f66a/plugin/markdown/markdown.js#L245. Though, I need to dig further as the `markdown.js` library is using `alert` to report error messages.

**Comment by vernhart** [comment]
A quick search leads me to phantomjs's onAlert handler which seems like a quick way to convert alerts to console.log messages: http://phantomjs.org/api/webpage/handler/on-alert.html

**Comment by astefanutti** [comment]
It happens that the `XMLHttpRequest` status is actually `0` and that the returned `responseText` is empty. I've isolated the smallest piece of code that instantiates a `XMLHttpRequest` and sends a `GET` request:

```
<html>
    <body>
        <script>
            var xhr = new XMLHttpRequest();
            xhr.onreadystatechange = function() {
                if( xhr.readyState === 4 ) {
                    if ( (xhr.status >= 200 && xhr.status < 300 ) || xhr.status === 0 ) {
                        // xhr.responseText.length == 0 !!!
                    }
                }
            };
            xhr.open( 'GET', "example.md", false );
            xhr.send();
        </script>
    </body>
</html>
```
And still face the empty response!

I don't see anything wrong in the above example. So I suspect this an issue at the level of PhantomJS though that'd be a pretty big one.

**Comment by astefanutti** [comment]
I've just been tested it with SlimerJS as I'm working on supporting it as another backend in addition to PhantomJS and the external markdown document gets loaded and rendered correctly. So definitely a PhantomJS issue (or Qt WebKit possibly). Still trying to find if that is an option missing or a bug...

**Comment by gavincarr** [comment]
Just wanted to add a note that the same issue appears to affect remark.js as well - including the markdown inline in a text area works fine, but including via a separate markdown file using remark's sourceUrl results in a blank pdf.

**Comment by amouat** [comment]
Is there a workaround for this? Do I have to copy my markdown into the HTML file or can I change to slimerJS?

**Comment by astefanutti** [comment]
@amouat, I still need to identify to root cause of this issue that only happens when loading external file with `XMLHttpRequest`. In the meantime, inlining the markdown works but I agree that's far from ideal. I'll investigate further on this ASAP.

**Comment by dylangmiles** [comment]
@astefantutti, I got it working. The `XMLHttpRequest.open` call in the markdown.js function passes false for the async parameter. On return to the `onreadystatechanged` callback the status is zero and the response text is empty. I suspect that phantomjs is not supporting synchronous requests??

Anyway, I then passed in false to the open call which loads the markdown. But the convertSlides function requires the processSlides to have completed which it has not because the call to XMLHttpRequest is now asynchronous. So I added a callback parameter to processSlides. This has some other knock issues  which I solved by refreshing Reveal.

Below are the changes I made:

```
Index: module/plugin/markdown/markdown.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- module/plugin/markdown/markdown.js	(revision 782a07b63b8f1472f148828fcf2780391564e157)
+++ module/plugin/markdown/markdown.js	(revision )
@@ -205,7 +205,7 @@
 	 * multi-slide markdown into separate sections and
 	 * handles loading of external markdown.
 	 */
-	function processSlides() {
+	function processSlides(complete) {
 
 		var sections = document.querySelectorAll( '[data-markdown]'),
 			section;
@@ -213,14 +213,12 @@
 		for( var i = 0, len = sections.length; i < len; i++ ) {
 
 			section = sections[i];
-
+			
 			if( section.getAttribute( 'data-markdown' ).length ) {
 
 				var xhr = new XMLHttpRequest(),
 					url = section.getAttribute( 'data-markdown' );
 				
-				console.log(url);
-
 				datacharset = section.getAttribute( 'data-charset' );
 
 				// see https://developer.mozilla.org/en-US/docs/Web/API/element.getAttribute#Notes
@@ -230,6 +228,7 @@
 
 				xhr.onreadystatechange = function() {
 					if( xhr.readyState === 4 ) {
+
 						// file protocol yields status code 0 (useful for local debug, mobile applications etc.)
 						if ( ( xhr.status >= 200 && xhr.status < 300 ) || xhr.status === 0 ) {
 
@@ -240,6 +239,8 @@
 								attributes: getForwardedAttributes( section )
 							});
 
+							complete(true);
+
 						}
 						else {
 
@@ -249,17 +250,20 @@
 								'<p>Remember that you need to serve the presentation HTML from a HTTP server.</p>' +
 								'</section>';
 
+							complete(false);
+
 						}
 					}
 				};
 
-				xhr.open( 'GET', url, false );
+				xhr.open( 'GET', url, true );
 
 				try {
 					xhr.send();
 				}
 				catch ( e ) {
 					alert( 'Failed to get the Markdown file ' + url + '. Make sure that the presentation and the file are served by a HTTP server and the file can be found there. ' + e );
+
 				}
 
 			}
@@ -272,9 +276,13 @@
 					attributes: getForwardedAttributes( section )
 				});
 
+				complete(true);
+
 			}
 			else {
 				section.innerHTML = createMarkdownSlide( getMarkdownFromSlide( section ) );
+
+				complete(true);
 			}
 		}
 
@@ -390,8 +398,14 @@
 	return {
 
 		initialize: function() {
-			processSlides();
+			processSlides(function(success) {
-			convertSlides();
+				convertSlides();
+				
+				//Refresh
+				Reveal.layout();
+				Reveal.slide(0);
+			});
+			
 		},
 
 		// TODO: Do these belong in the API?
```

**Comment by s-bernard** [comment]
@dylangmiles Your modifications make phantomjs crash on my system, I have developped another version which use xhr.onload to call the callback.


[0001-Set-asynchronous-load-of-external-markdown.txt](https://github.com/astefanutti/decktape/files/54381/0001-Set-asynchronous-load-of-external-markdown.txt)

**Comment by acdcjunior** [comment]
I'm facing the same problem (reveal.js+markdown) and am using @s-bernard's patch. It is working so far.

**Comment by acdcjunior** [comment]
@s-bernard I'm not 100% on this (because I changed so much more code :D), but I think your patch somehow makes the slides' background style stop working. Something like this:

    # SlideOne
    ---
    
    # SlideTwo

...won't make the second slide's bg color red, as it should. Changing the `Reveal.layout()` line to `Reveal.sync()` seems to fix it for me.

**Comment by s-bernard** [comment]
I changed the line but it still breaks revealjs tests: https://github.com/hakimel/reveal.js/pull/1458
And I am not good enough in JS to understand what really happens.

**Comment by astefanutti** [comment]
Issue ariya/phantomjs#13908 has just been created and confirms synchronous AJAX requests don't work in current version of PhantomJS. So we may have a fix for this soon hopefully.

**Comment by astefanutti** [comment]
So ariya/phantomjs#13908 has been fixed. I've just rebased DeckTape on ariya/phantomjs@3d9a327 and I confirm it works now!

So I'll be uploading updated binaries with the fix ASAP. In the meantime, you can pull latest DeckTape revision and build the PhantomJS submodule.

**Comment by acdcjunior** [comment]
Great, thanks @astefanutti!

**Comment by acdcjunior** [comment]
@astefanutti Just to confirm, are the binaries still the previous ones? I mean, to get the fix we have to build the PhantomJS submodule ourselves for now, right?

**Comment by astefanutti** [comment]
@acdcjunior your understanding is correct. I'll update the binaries ASAP though I'm waiting to have a couple of other fixes before.

**Comment by andrewhao** [comment]
Can someone confirm if their Docker build is working for external Markdown files? Mine doesn't seem to be doing so (I'm on remark.js).

**Comment by astefanutti** [comment]
@andrewhao, the current Docker image does not contain the fix yet. I'm waiting for another issue's fix and I'll update the image ASAP then.

**Comment by lornajane** [comment]
I'm having exactly this problem with the docker image producing blank output for my reveal.js presentation which has markdown in a separate file.  Let me know if I can lend a hand testing a new docker image or something.

**Comment by astefanutti** [comment]
@lornajane I'll update the Docker image ASAP and keep you in touch. Just in case, the pre-compiled Mac OS and Ubuntu binaries should already contain the fix.

**Comment by pmurkowsky** [comment]
@astefanutti I saw that you updated the docker image 15 days ago, but it does not seem to contain this fix yet. Is this correct?

**Comment by astefanutti** [comment]
@pmurkowsky your understanding is correct. I'm working on another issue and should be able to update the Docker image containing both fixes soon.

**Comment by astefanutti** [comment]
I've just released version `1.0.0`, which contains the updated binaries with the fix for this issue. The Docker image has been updated as well.

Thanks for your patience!
